### PR TITLE
Jetpack connect: Make set reducer pure

### PR DESCRIPTION
**Pull request by sirreal**
This PR moves impure `Date.now()` out of a reducer and into `dispatch`. This was breaking the purity convention for reducers.

There should be no functional changes.

## Testing
1. Logged in test flow should continue to work properly.
1. After entering logged-in auth flow starting from WP Admin, inspect the state tree to ensure that the value is populated and appears correct:
   ```js
   new Date( state.jetpackConnect.jetpackConnectAuthorize.timestamp )
   ```

### CODE REVIEW

File: client/state/jetpack-connect/actions.js [hunk 1]
```diff
@@ -16,31 +16,31 @@ import wpcom from 'lib/wp';
...
 	JETPACK_CONNECT_SSO_VALIDATION_REQUEST,
 	JETPACK_CONNECT_SSO_VALIDATION_SUCCESS,
-	JETPACK_CONNECT_SSO_VALIDATION_ERROR,
 	JETPACK_CONNECT_USER_ALREADY_CONNECTED,
-	SITES_RECEIVE,
 	SITE_REQUEST,
-	SITE_REQUEST_SUCCESS,
 	SITE_REQUEST_FAILURE,
+	SITE_REQUEST_SUCCESS,
+	SITES_RECEIVE,
```
Comment (sirreal) [inline client/state/jetpack-connect/actions.js]: Just 🔤

File: client/jetpack-connect/controller.js [hunk 2]
```diff
@@ -180,15 +180,17 @@ export function signupForm( context, next ) {
...
+		// However, clientId is required for some reducer logic :(
+		//
+		// Set timestamp here so reducer can remain pure
 		//
 		// Hopefully when actions move to data-layer, this will become clearer and
 		// we won't need to store clientId in state
 		//
 		context.store.dispatch( {
 			type: JETPACK_CONNECT_QUERY_SET,
-			from: transformedQuery.from,
```
Comment (sirreal) [inline client/jetpack-connect/controller.js]: This requirement was removed in #20859. Clean up here.

File: client/jetpack-connect/controller.js [hunk 3]
```diff
@@ -180,15 +180,17 @@ export function signupForm( context, next ) {
...
 			clientId: transformedQuery.clientId,
+			timestamp: Date.now(),
```
Comment (sirreal) [inline client/jetpack-connect/controller.js]: A dedicated action creator could wrap this, but for now just set it here. This state may still evolve significantly.

File: client/state/jetpack-connect/reducer/jetpack-connect-authorize.js [hunk 4]
```diff
@@ -67,10 +67,10 @@ export default function jetpackConnectAuthorize( state = {}, action ) {
 			return {
 				authorizeError: false,
 				authorizeSuccess: false,
+				clientId: action.clientId,
 				isAuthorizing: false,
-				timestamp: Date.now(),
+				timestamp: action.timestamp,
```
Comment (sirreal) [inline client/state/jetpack-connect/reducer/jetpack-connect-authorize.js]: Reducer purity 🎉

**Comment by ockham** [comment]
> 1. After entering logged-in auth flow starting from WP Admin, inspect the state tree to ensure that the value is populated and appears correct:
>   ```js
>   new Date( state.jetpackConnect.jetpackConnectAuthorize.timestamp )
>   ```
  
Err, how do I force my site to redirect to dev (calypso.localhost:3000) instead of wp.com?

**Comment by tyxla** [comment]
> Err, how do I force my site to redirect to dev (calypso.localhost:3000) instead of wp.com?

By adding `&calypso_env=development` to the connection URL that you get redirected to when you click the Connect button in wp-admin.

**Comment by seear** [comment]
Tests fine for me. I can't find where this timestamp is used?

**Comment by sirreal** [comment]
> Tests fine for me. I can't find where this timestamp is used?

Checking `isStale`:

https://github.com/Automattic/wp-calypso/blob/47d9f87fe8e141795f85d17dfedb9ae33e57abe5/client/state/jetpack-connect/reducer/jetpack-connect-authorize.js#L119

This pattern appears several places around Jetpack Connect.

**Pull request merged**